Background-Substractor-Video-Capturer/main.cpp

```diff
@@ -0,0 +1,328 @@
+static std::atomic<bool> g_is_publishing(false);
+
+struct custom_video_capturer {
+  const otc_video_capturer *video_capturer;
+  struct otc_video_capturer_callbacks video_capturer_callbacks;
+  int format;
+  int width;
+  int height;
+  int fps;
+  otk_thread_t capturer_thread;
+  std::atomic<bool> capturer_thread_exit;
+};
+
+static otk_thread_func_return_type capturer_thread_start_function(void *arg) {
+  struct custom_video_capturer *video_capturer = static_cast<struct custom_video_capturer *>(arg);
+  if (video_capturer == nullptr) {
+    otk_thread_func_return_value;
+  }
+
+  cv::VideoCapture opencv_video_capture(0);
+  if (!opencv_video_capture.isOpened()) {
+    otk_thread_func_return_value;
+  }
+
+  // The following arguments have the default values according to
+  // https://docs.opencv.org/master/de/de1/group__video__motion.html.
+  int history = 500;
+  double var_threshold = 16;
+  bool detect_shadows = true;
+  cv::Ptr<cv::BackgroundSubtractor> background_subtractor =
+      cv::createBackgroundSubtractorMOG2(history, var_threshold, detect_shadows);
+  cv::Mat opencv_frame, foreground_mask, background_image, foreground_image;
+  while((video_capturer->capturer_thread_exit.load() == false) &&
+        opencv_video_capture.read(opencv_frame)) {
+    if (foreground_mask.empty()) {
+      foreground_mask.create(opencv_frame.size(), opencv_frame.type());
+    }
+    // See https://docs.opencv.org/master/d7/d7b/classcv_1_1BackgroundSubtractorMOG2.html
+    // for more info about the learing rate argument.
+    background_subtractor->apply(opencv_frame, foreground_mask, /*learning_rate*/ -1);
+    // Smoothing the image for the mask with GaussianBlur() function.
+    // See https://docs.opencv.org/master/dc/dd3/tutorial_gausian_median_blur_bilateral_filter.html.
+    cv::GaussianBlur(foreground_mask, foreground_mask, cv::Size(11, 11), 3.5, 3.5);
+    // Threshold the image for the mask too.
+    // See https://docs.opencv.org/master/d7/d4d/tutorial_py_thresholding.html
+    cv::threshold(foreground_mask, foreground_mask, 10, 255, cv::THRESH_BINARY);
+    // New black foreground image.
+    foreground_image = cv::Scalar::all(0);
+    opencv_frame.copyTo(foreground_image, foreground_mask);
+
+    cv::cvtColor(foreground_image, foreground_image, cv::COLOR_BGR2YUV);
+    std::vector<cv::Mat> yuv;
+    cv::split(foreground_image, yuv);
+    // Convert cv::Mat video frame to otc_video_frame video frame.
+    otc_video_frame *otc_frame = otc_video_frame_new_I420(
+        video_capturer->width,
+        video_capturer->height,
+        yuv[0].ptr<uchar>(),
+        yuv[0].step[0],
+        yuv[1].ptr<uchar>(),
+        yuv[1].step[0],
+        yuv[2].ptr<uchar>(),
+        yuv[2].step[0]);
+    otc_video_capturer_provide_frame(video_capturer->video_capturer, 0, otc_frame);
+    if (otc_frame != nullptr) {
+      otc_video_frame_delete(otc_frame);
+    }
+  }
+
+  opencv_video_capture.release();
+
+  otk_thread_func_return_value;
+}
+
```

Comment:
Looks like there's really two parts to this sample app: one part (above) that uses OpenCV to perform background subtraction; and another part (below) that wires it up w/ OpenTok.  How do you feel about abstracting the open to another module, since it seems like we're mixing concerns.